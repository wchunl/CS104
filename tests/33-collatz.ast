ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_FUNCTION "" (7.9.1) function
|   |   TOK_TYPE_ID "" (7.9.1) typeid
|   |   |   TOK_INT "int" (7.9.1) int
|   |   |   TOK_IDENT "collatz" (7.9.5) variable
|   |   TOK_PARAM "(" (7.9.13) param
|   |   |   TOK_TYPE_ID "" (7.9.14) typeid
|   |   |   |   TOK_INT "int" (7.9.14) int
|   |   |   |   TOK_IDENT "n" (7.9.18) int variable (7.9.14)
|   |   TOK_BLOCK "{" (7.9.21)
|   |   |   TOK_VARDECL "=" (7.10.10) variable lval
|   |   |   |   TOK_INT "int" (7.10.4) int
|   |   |   |   TOK_IDENT "c" (7.10.8) int variable (7.10.10)
|   |   |   |   TOK_INTCON "0" (7.10.12) int const
|   |   |   TOK_WHILE "while" (7.11.4)
|   |   |   |   TOK_NE "!=" (7.11.13)
|   |   |   |   |   TOK_IDENT "n" (7.11.11) int variable (7.9.14)
|   |   |   |   |   TOK_INTCON "1" (7.11.16) int const
|   |   |   |   TOK_BLOCK "{" (7.11.19)
|   |   |   |   |   '=' "=" (7.12.9) int vreg
|   |   |   |   |   |   TOK_IDENT "c" (7.12.7) int variable (7.10.10)
|   |   |   |   |   |   '+' "+" (7.12.13) int vreg
|   |   |   |   |   |   |   TOK_IDENT "c" (7.12.11) int variable (7.10.10)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.12.15) int const
|   |   |   |   |   TOK_IF "if" (7.13.7)
|   |   |   |   |   |   TOK_EQ "==" (7.13.17) vreg
|   |   |   |   |   |   |   '%' "%" (7.13.13) int vreg
|   |   |   |   |   |   |   |   TOK_IDENT "n" (7.13.11) int variable (7.9.14)
|   |   |   |   |   |   |   |   TOK_INTCON "2" (7.13.15) int const
|   |   |   |   |   |   |   TOK_INTCON "1" (7.13.20) int const
|   |   |   |   |   |   '=' "=" (7.13.25) int vreg
|   |   |   |   |   |   |   TOK_IDENT "n" (7.13.23) int variable (7.9.14)
|   |   |   |   |   |   |   '+' "+" (7.13.33) int vreg
|   |   |   |   |   |   |   |   '*' "*" (7.13.29) int vreg
|   |   |   |   |   |   |   |   |   TOK_INTCON "3" (7.13.27) int const
|   |   |   |   |   |   |   |   |   TOK_IDENT "n" (7.13.31) int variable (7.9.14)
|   |   |   |   |   |   |   |   TOK_INTCON "1" (7.13.35) int const
|   |   |   |   |   |   TOK_IFELSE "=" (7.14.25)
|   |   |   |   |   |   |   TOK_IDENT "n" (7.14.23) int variable (7.9.14)
|   |   |   |   |   |   |   '/' "/" (7.14.29) int vreg
|   |   |   |   |   |   |   |   TOK_IDENT "n" (7.14.27) int variable (7.9.14)
|   |   |   |   |   |   |   |   TOK_INTCON "2" (7.14.31) int const
|   |   |   TOK_RETURN "return" (7.16.4)
|   |   |   |   TOK_IDENT "c" (7.16.11) int variable (7.10.10)
|   TOK_FUNCTION "" (7.19.1) function
|   |   TOK_TYPE_ID "" (7.19.1) typeid
|   |   |   TOK_VOID "void" (7.19.1) void
|   |   |   TOK_IDENT "test" (7.19.6) variable
|   |   TOK_PARAM "(" (7.19.11) param
|   |   |   TOK_TYPE_ID "" (7.19.12) typeid
|   |   |   |   TOK_INT "int" (7.19.12) int
|   |   |   |   TOK_IDENT "n" (7.19.16) int variable (7.9.14)
|   |   TOK_BLOCK "{" (7.19.19)
|   |   |   TOK_VARDECL "=" (7.20.10) variable lval
|   |   |   |   TOK_INT "int" (7.20.4) int
|   |   |   |   TOK_IDENT "c" (7.20.8) int variable (7.10.10)
|   |   |   |   TOK_CALL "(" (7.20.20)
|   |   |   |   |   TOK_IDENT "collatz" (7.20.12) variable
|   |   |   |   |   TOK_IDENT "n" (7.20.21) int variable (7.9.14)
|   |   |   TOK_CALL "(" (7.21.11)
|   |   |   |   TOK_IDENT "putstr" (7.21.4) variable
|   |   |   |   TOK_STRINGCON ""collatz("" (7.21.12) string const
|   |   |   TOK_CALL "(" (7.22.11)
|   |   |   |   TOK_IDENT "putint" (7.22.4) variable
|   |   |   |   TOK_IDENT "n" (7.22.12) int variable (7.9.14)
|   |   |   TOK_CALL "(" (7.23.11)
|   |   |   |   TOK_IDENT "putstr" (7.23.4) variable
|   |   |   |   TOK_STRINGCON "") = "" (7.23.12) string const
|   |   |   TOK_CALL "(" (7.24.11)
|   |   |   |   TOK_IDENT "putint" (7.24.4) variable
|   |   |   |   TOK_IDENT "c" (7.24.12) int variable (7.10.10)
|   |   |   TOK_CALL "(" (7.25.11)
|   |   |   |   TOK_IDENT "putchr" (7.25.4) variable
|   |   |   |   TOK_CHARCON "'\n'" (7.25.12) int const
|   TOK_FUNCTION "" (7.28.1) function
|   |   TOK_TYPE_ID "" (7.28.1) typeid
|   |   |   TOK_INT "int" (7.28.1) int
|   |   |   TOK_IDENT "main" (7.28.5) variable
|   |   TOK_PARAM "(" (7.28.9) param
|   |   TOK_BLOCK "{" (7.28.12)
|   |   |   TOK_CALL "(" (7.29.9)
|   |   |   |   TOK_IDENT "test" (7.29.4) variable
|   |   |   |   TOK_INTCON "3" (7.29.10) int const
|   |   |   TOK_CALL "(" (7.30.9)
|   |   |   |   TOK_IDENT "test" (7.30.4) variable
|   |   |   |   TOK_INTCON "50" (7.30.10) int const
|   |   |   TOK_RETURN "return" (7.31.4)
|   |   |   |   TOK_INTCON "0" (7.31.11) int const
