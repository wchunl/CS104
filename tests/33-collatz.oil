fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
collatz:  .function int
          .param int n
          .local int c
          c = 0
.wh0:     $t0:i = n != 1
          goto .od0 if not $t0:i
.do0:     $t1:i = c + 1
          c = $t1:i
.if0:     $t2:i = % == 1
          goto .el0 if not $t2:i
.th0:     $t3:i = * + 1
          n = $t3:i
.el0:     $t4:i = n / 2
          n = $t4:i
.fi0:     
.od0:     return c
          return 
          .end
test:     .function
          .param int n
          .local int c
          collatz(n)
          c = $t5:i
          putstr("collatz(")
          putint(n)
          putstr(") = ")
          putint(c)
          putchr('\n')
          .end
main:     .function int
          test(3)
          test(50)
          return 0
          return 
          .end
