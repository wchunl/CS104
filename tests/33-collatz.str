string_set[   2]:   12597418120160561823 0xc22d38->"int"
string_set[   5]:    1370730858159036685 0xc25288->"+"
string_set[   7]:   10959529184379665549 0xc24d58->"c"
string_set[   8]:   13497578857207247412 0xc22bd8->"string"
                    4551451650890805270 0xc25818->"2"
string_set[   9]:   10319632053171904911 0xc24f68->"while"
string_set[  10]:    9201800968376702500 0xc23fc8->"getchr"
string_set[  14]:   10462659240778295294 0xc24738->"collatz"
                   15816506932648033945 0xc22b28->"("
string_set[  15]:   12868445110721718657 0xc22f38->")"
string_set[  17]:    6142509188972423790 0xc1e778->""
string_set[  20]:    1583576014053361350 0xc23b98->"str"
string_set[  21]:    9317282556992397945 0xc255f8->"if"
string_set[  22]:   10159970873491820195 0xc251d8->"1"
string_set[  26]:   12291392672316406109 0xc230f8->"line"
                    9611428142580219387 0xc1e798->"fail"
                   11597697714117577063 0xc24e08->"="
string_set[  28]:   18413295556186398326 0xc23308->"putint"
string_set[  29]:   11911796782241914524 0xc1e318->"void"
string_set[  35]:   11868120863286193964 0xc26fc8->"main"
string_set[  39]:    7419612306637001510 0xc26998->"") = ""
string_set[  40]:   12137009137744407700 0xc27518->"50"
string_set[  42]:   13686721134187162920 0xc25598->"}"
string_set[  49]:   12679825168387489523 0xc239e8->"num"
string_set[  50]:      21606623943142546 0xc232b8->"{"
string_set[  52]:    5344662657899890615 0xc23118->";"
string_set[  54]:    7296690267450279874 0xc235c8->"chr"
string_set[  55]:    2297668033614959926 0xc24eb8->"0"
string_set[  59]:    6192799629375248351 0xc267f8->"'\n'"
string_set[  60]:    1873634190020599143 0xc25cf8->"else"
string_set[  61]:   12865648880272446602 0xc25fe8->"/"
string_set[  62]:     821927901342812852 0xc267b8->""collatz(""
string_set[  63]:    9826907452181669771 0xc249d8->"n"
string_set[  64]:   16579916199511856337 0xc244c8->"getln"
string_set[  67]:    4706206694200543916 0xc22e08->","
string_set[  73]:    9386170080689500517 0xc22cb8->"expr"
string_set[  75]:    8248777770799913213 0xc258e8->"3"
string_set[  76]:   11166701219533395393 0xc24238->"getstr"
string_set[  80]:    7211130878751977441 0xc23458->"putchr"
                    2617056156435215158 0xc25128->"!="
string_set[  84]:    6608372942718979652 0xc25768->"%"
string_set[  85]:    4017852018407011111 0xc22f18->"file"
string_set[  87]:   14635384904416252381 0xc24f88->"return"
string_set[  89]:   13348923631108115385 0xc23778->"putstr"
                    6637313742931709005 0xc25be8->"*"
string_set[  94]:   15118982290295364091 0xc26298->"test"
string_set[  95]:   14736352883882276958 0xc258c8->"=="
load_factor = 0.485
bucket_count = 97
max_bucket_size = 3
