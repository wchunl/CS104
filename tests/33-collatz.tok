#  1 "33-collatz.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "33-collatz.oc"
#  1 "oclib.h"
# 22 "oclib.h"
   6 22.00001 261 TOK_VOID      void
   6 22.00006 282 TOK_IDENT     fail
   6 22.00011  40 '('           (
   6 22.00012 263 TOK_STRING    string
   6 22.00019 282 TOK_IDENT     expr
   6 22.00023  44 ','           ,
   6 22.00025 263 TOK_STRING    string
   6 22.00032 282 TOK_IDENT     file
   6 22.00036  44 ','           ,
   6 22.00038 262 TOK_INT       int
   6 22.00042 282 TOK_IDENT     line
   6 22.00046  41 ')'           )
   6 22.00047  59 ';'           ;
   6 24.00001 261 TOK_VOID      void
   6 24.00006 282 TOK_IDENT     putchr
   6 24.00013  40 '('           (
   6 24.00014 262 TOK_INT       int
   6 24.00018 282 TOK_IDENT     chr
   6 24.00021  41 ')'           )
   6 24.00022  59 ';'           ;
   6 25.00001 261 TOK_VOID      void
   6 25.00006 282 TOK_IDENT     putint
   6 25.00013  40 '('           (
   6 25.00014 262 TOK_INT       int
   6 25.00018 282 TOK_IDENT     num
   6 25.00021  41 ')'           )
   6 25.00022  59 ';'           ;
   6 26.00001 261 TOK_VOID      void
   6 26.00006 282 TOK_IDENT     putstr
   6 26.00013  40 '('           (
   6 26.00014 263 TOK_STRING    string
   6 26.00021 282 TOK_IDENT     str
   6 26.00024  41 ')'           )
   6 26.00025  59 ';'           ;
   6 28.00001 262 TOK_INT       int
   6 28.00005 282 TOK_IDENT     getchr
   6 28.00011  40 '('           (
   6 28.00012  41 ')'           )
   6 28.00013  59 ';'           ;
   6 29.00001 263 TOK_STRING    string
   6 29.00008 282 TOK_IDENT     getstr
   6 29.00014  40 '('           (
   6 29.00015  41 ')'           )
   6 29.00016  59 ';'           ;
   6 30.00001 263 TOK_STRING    string
   6 30.00008 282 TOK_IDENT     getln
   6 30.00013  40 '('           (
   6 30.00014  41 ')'           )
   6 30.00015  59 ';'           ;
#  8 "33-collatz.oc"
   7 9.00001 262 TOK_INT       int
   7 9.00005 282 TOK_IDENT     collatz
   7 9.00013  40 '('           (
   7 9.00014 262 TOK_INT       int
   7 9.00018 282 TOK_IDENT     n
   7 9.00019  41 ')'           )
   7 9.00021 123 '{'           {
   7 10.00004 262 TOK_INT       int
   7 10.00008 282 TOK_IDENT     c
   7 10.00010  61 '='           =
   7 10.00012 283 TOK_INTCON    0
   7 10.00013  59 ';'           ;
   7 11.00004 267 TOK_WHILE     while
   7 11.00010  40 '('           (
   7 11.00011 282 TOK_IDENT     n
   7 11.00013 276 TOK_NE        !=
   7 11.00016 283 TOK_INTCON    1
   7 11.00017  41 ')'           )
   7 11.00019 123 '{'           {
   7 12.00007 282 TOK_IDENT     c
   7 12.00009  61 '='           =
   7 12.00011 282 TOK_IDENT     c
   7 12.00013  43 '+'           +
   7 12.00015 283 TOK_INTCON    1
   7 12.00016  59 ';'           ;
   7 13.00007 265 TOK_IF        if
   7 13.00010  40 '('           (
   7 13.00011 282 TOK_IDENT     n
   7 13.00013  37 '%'           %
   7 13.00015 283 TOK_INTCON    2
   7 13.00017 275 TOK_EQ        ==
   7 13.00020 283 TOK_INTCON    1
   7 13.00021  41 ')'           )
   7 13.00023 282 TOK_IDENT     n
   7 13.00025  61 '='           =
   7 13.00027 283 TOK_INTCON    3
   7 13.00029  42 '*'           *
   7 13.00031 282 TOK_IDENT     n
   7 13.00033  43 '+'           +
   7 13.00035 283 TOK_INTCON    1
   7 13.00036  59 ';'           ;
   7 14.00018 266 TOK_ELSE      else
   7 14.00023 282 TOK_IDENT     n
   7 14.00025  61 '='           =
   7 14.00027 282 TOK_IDENT     n
   7 14.00029  47 '/'           /
   7 14.00031 283 TOK_INTCON    2
   7 14.00032  59 ';'           ;
   7 15.00004 125 '}'           }
   7 16.00004 268 TOK_RETURN    return
   7 16.00011 282 TOK_IDENT     c
   7 16.00012  59 ';'           ;
   7 17.00001 125 '}'           }
   7 19.00001 261 TOK_VOID      void
   7 19.00006 282 TOK_IDENT     test
   7 19.00011  40 '('           (
   7 19.00012 262 TOK_INT       int
   7 19.00016 282 TOK_IDENT     n
   7 19.00017  41 ')'           )
   7 19.00019 123 '{'           {
   7 20.00004 262 TOK_INT       int
   7 20.00008 282 TOK_IDENT     c
   7 20.00010  61 '='           =
   7 20.00012 282 TOK_IDENT     collatz
   7 20.00020  40 '('           (
   7 20.00021 282 TOK_IDENT     n
   7 20.00022  41 ')'           )
   7 20.00023  59 ';'           ;
   7 21.00004 282 TOK_IDENT     putstr
   7 21.00011  40 '('           (
   7 21.00012 285 TOK_STRINGCON "collatz("
   7 21.00022  41 ')'           )
   7 21.00023  59 ';'           ;
   7 22.00004 282 TOK_IDENT     putint
   7 22.00011  40 '('           (
   7 22.00012 282 TOK_IDENT     n
   7 22.00013  41 ')'           )
   7 22.00014  59 ';'           ;
   7 23.00004 282 TOK_IDENT     putstr
   7 23.00011  40 '('           (
   7 23.00012 285 TOK_STRINGCON ") = "
   7 23.00018  41 ')'           )
   7 23.00019  59 ';'           ;
   7 24.00004 282 TOK_IDENT     putint
   7 24.00011  40 '('           (
   7 24.00012 282 TOK_IDENT     c
   7 24.00013  41 ')'           )
   7 24.00014  59 ';'           ;
   7 25.00004 282 TOK_IDENT     putchr
   7 25.00011  40 '('           (
   7 25.00012 284 TOK_CHARCON   '\n'
   7 25.00016  41 ')'           )
   7 25.00017  59 ';'           ;
   7 26.00001 125 '}'           }
   7 28.00001 262 TOK_INT       int
   7 28.00005 282 TOK_IDENT     main
   7 28.00009  40 '('           (
   7 28.00010  41 ')'           )
   7 28.00012 123 '{'           {
   7 29.00004 282 TOK_IDENT     test
   7 29.00009  40 '('           (
   7 29.00010 283 TOK_INTCON    3
   7 29.00011  41 ')'           )
   7 29.00012  59 ';'           ;
   7 30.00004 282 TOK_IDENT     test
   7 30.00009  40 '('           (
   7 30.00010 283 TOK_INTCON    50
   7 30.00012  41 ')'           )
   7 30.00013  59 ';'           ;
   7 31.00004 268 TOK_RETURN    return
   7 31.00011 283 TOK_INTCON    0
   7 31.00012  59 ';'           ;
   7 32.00001 125 '}'           }
