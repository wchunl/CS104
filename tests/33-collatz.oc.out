; 0.  1: # 1 "33-collatz.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "33-collatz.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: 
; 4.  4: 
; 4.  5: 
; 4.  6: 
; 4.  7: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 8 "33-collatz.oc" 2
; 7.  8: 
; 7.  9: int collatz (int n) {
; 7. 10:    int c = 0;
; 7. 11:    while (n != 1) {
; 7. 12:       c = c + 1;
; 7. 13:       if (n % 2 == 1) n = 3 * n + 1;
; 7. 14:                  else n = n / 2;
; 7. 15:    }
; 7. 16:    return c;
; 7. 17: }
; 7. 18: 
; 7. 19: void test (int n) {
; 7. 20:    int c = collatz (n);
; 7. 21:    putstr ("collatz(");
; 7. 22:    putint (n);
; 7. 23:    putstr (") = ");
; 7. 24:    putint (c);
; 7. 25:    putchr ('\n');
; 7. 26: }
; 7. 27: 
; 7. 28: int main() {
; 7. 29:    test (3);
; 7. 30:    test (50);
; 7. 31:    return 0;
; 7. 32: }
